File: apps/server/src/core/security.ts
// Module: Provides password hashing and JWT signing helpers for authentication.
import {
  createHmac,
  pbkdf2Sync,
  randomBytes,
  timingSafeEqual,
} from "node:crypto"

import type { Settings } from "./config"

const JWT_ALGORITHM = "HS256"
const PASSWORD_ALGORITHM = "pbkdf2_sha256"
const PASSWORD_ITERATIONS = 200_000

type JwtPayload = {
  sub: string
  exp: number
  iat: number
}

function base64urlEncode(value: Buffer): string {
  return value.toString("base64url")
}

function base64urlDecode(value: string): Buffer {
  return Buffer.from(value, "base64url")
}

export function hashPassword(password: string): string {
  const salt = randomBytes(16)
  const passwordHash = pbkdf2Sync(
    password,
    salt,
    PASSWORD_ITERATIONS,
    32,
    "sha256"
  )

  return [
    PASSWORD_ALGORITHM,
    PASSWORD_ITERATIONS.toString(),
    base64urlEncode(salt),
    base64urlEncode(passwordHash),
  ].join("$")
}

export function verifyPassword(
  password: string,
  storedPasswordHash: string
): boolean {
  const parts = storedPasswordHash.split("$")
  if (parts.length !== 4) {
    return false
  }

  const [algorithm, iterationsText, saltText, expectedHashText] = parts
  if (algorithm !== PASSWORD_ALGORITHM) {
    return false
  }

  const iterations = Number.parseInt(iterationsText, 10)
  if (!Number.isFinite(iterations)) {
    return false
  }

  let salt: Buffer
  let expectedHash: Buffer

  try {
    salt = base64urlDecode(saltText)
    expectedHash = base64urlDecode(expectedHashText)
  } catch {
    return false
  }

  const passwordHash = pbkdf2Sync(
    password,
    salt,
    iterations,
    expectedHash.length,
    "sha256"
  )

  if (passwordHash.length !== expectedHash.length) {
    return false
  }

  return timingSafeEqual(passwordHash, expectedHash)
}

export function createAccessToken(
  subject: string,
  settings: Settings,
  expiresInMinutes = settings.accessTokenExpireMinutes
): string {
  const issuedAt = Math.floor(Date.now() / 1000)
  const expiresAt = issuedAt + expiresInMinutes * 60
  const header = {
    alg: JWT_ALGORITHM,
    typ: "JWT",
  }
  const payload: JwtPayload = {
    sub: subject,
    exp: expiresAt,
    iat: issuedAt,
  }

  const encodedHeader = base64urlEncode(
    Buffer.from(JSON.stringify(header), "utf8")
  )
  const encodedPayload = base64urlEncode(
    Buffer.from(JSON.stringify(payload), "utf8")
  )
  const signingInput = `${encodedHeader}.${encodedPayload}`
  const signature = createHmac("sha256", settings.accessTokenSecret)
    .update(signingInput)
    .digest()

  return `${signingInput}.${base64urlEncode(signature)}`
}

export function decodeAccessToken(
  token: string,
  settings: Settings
): JwtPayload {
  const [encodedHeader, encodedPayload, encodedSignature] = token.split(".")
  if (!encodedHeader || !encodedPayload || !encodedSignature) {
    throw new Error("Invalid token")
  }

  const signingInput = `${encodedHeader}.${encodedPayload}`
  const expectedSignature = createHmac("sha256", settings.accessTokenSecret)
    .update(signingInput)
    .digest()
  const providedSignature = base64urlDecode(encodedSignature)

  if (
    providedSignature.length !== expectedSignature.length ||
    !timingSafeEqual(expectedSignature, providedSignature)
  ) {
    throw new Error("Invalid token signature")
  }

  const header = JSON.parse(
    base64urlDecode(encodedHeader).toString("utf8")
  ) as {
    alg?: unknown
  }
  if (header.alg !== JWT_ALGORITHM) {
    throw new Error("Invalid token algorithm")
  }

  const payload = JSON.parse(
    base64urlDecode(encodedPayload).toString("utf8")
  ) as Partial<JwtPayload>

  if (typeof payload.sub !== "string" || payload.sub.length === 0) {
    throw new Error("Invalid token subject")
  }

  if (typeof payload.exp !== "number" || Date.now() / 1000 >= payload.exp) {
    throw new Error("Expired token")
  }

  if (typeof payload.iat !== "number") {
    throw new Error("Invalid token issued time")
  }

  return {
    sub: payload.sub,
    exp: payload.exp,
    iat: payload.iat,
  }
}

File: apps/server/app/core/security.py
# Module: Provides password hashing and JWT signing helpers for authentication.
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from datetime import UTC, datetime, timedelta
from os import urandom
from typing import Any

from app.core.config import get_settings

JWT_ALGORITHM = "HS256"
PASSWORD_ALGORITHM = "pbkdf2_sha256"
PASSWORD_ITERATIONS = 200_000


def _base64url_encode(value: bytes) -> str:
    return base64.urlsafe_b64encode(value).rstrip(b"=").decode("ascii")


def _base64url_decode(value: str) -> bytes:
    padded_value = value + "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(padded_value.encode("ascii"))


def hash_password(password: str) -> str:
    salt = urandom(16)
    password_hash = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        salt,
        PASSWORD_ITERATIONS,
    )

    return "$".join(
        [
            PASSWORD_ALGORITHM,
            str(PASSWORD_ITERATIONS),
            _base64url_encode(salt),
            _base64url_encode(password_hash),
        ]
    )


def verify_password(password: str, stored_password_hash: str) -> bool:
    try:
        algorithm, iterations_text, salt_text, expected_hash_text = stored_password_hash.split("$")
        if algorithm != PASSWORD_ALGORITHM:
            return False

        iterations = int(iterations_text)
        salt = _base64url_decode(salt_text)
        expected_hash = _base64url_decode(expected_hash_text)
    except (ValueError, TypeError):
        return False

    password_hash = hashlib.pbkdf2_hmac(
        "sha256",
        password.encode("utf-8"),
        salt,
        iterations,
    )

    return hmac.compare_digest(password_hash, expected_hash)


def create_access_token(subject: str, expires_delta: timedelta | None = None) -> str:
    settings = get_settings()
    expires_at = datetime.now(UTC) + (
        expires_delta or timedelta(minutes=settings.access_token_expire_minutes)
    )
    header = {
        "alg": JWT_ALGORITHM,
        "typ": "JWT",
    }
    payload = {
        "sub": subject,
        "exp": int(expires_at.timestamp()),
        "iat": int(datetime.now(UTC).timestamp()),
    }

    encoded_header = _base64url_encode(
        json.dumps(header, separators=(",", ":")).encode("utf-8")
    )
    encoded_payload = _base64url_encode(
        json.dumps(payload, separators=(",", ":")).encode("utf-8")
    )
    signing_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
    signature = hmac.new(
        settings.access_token_secret.encode("utf-8"),
        signing_input,
        hashlib.sha256,
    ).digest()

    return f"{encoded_header}.{encoded_payload}.{_base64url_encode(signature)}"


def decode_access_token(token: str) -> dict[str, Any]:
    settings = get_settings()

    try:
        encoded_header, encoded_payload, encoded_signature = token.split(".")
        signing_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
        expected_signature = hmac.new(
            settings.access_token_secret.encode("utf-8"),
            signing_input,
            hashlib.sha256,
        ).digest()
        provided_signature = _base64url_decode(encoded_signature)
        header = json.loads(_base64url_decode(encoded_header))
        payload = json.loads(_base64url_decode(encoded_payload))
    except (ValueError, json.JSONDecodeError):
        raise ValueError("Invalid token") from None

    if not hmac.compare_digest(expected_signature, provided_signature):
        raise ValueError("Invalid token signature")

    if header.get("alg") != JWT_ALGORITHM:
        raise ValueError("Invalid token algorithm")

    expires_at = payload.get("exp")
    if not isinstance(expires_at, int):
        raise ValueError("Invalid token expiry")

    if datetime.now(UTC).timestamp() >= expires_at:
        raise ValueError("Expired token")

    subject = payload.get("sub")
    if not isinstance(subject, str) or not subject:
        raise ValueError("Invalid token subject")

    return payload
